### DISCUSSION TITLE
How to do program analysis on top of tree sitter?

### DISCUSSION THREAD

**Discussion by smith-co** [tree-sitter/tree-sitter#2941] [Q&A]
I am curious to know whether is it possible to do data-flow analysis on top of tree-sitter?

Anyone aware of any such implementation?

Please help.

**Comment by NicholasLYang** [comment]
That's certainly possible although I'm not aware of any implementations. You'd probably need to take the tree-sitter tree and convert it into an IR, and then do regular data flow analysis on that, since a CST with a uniform `Node` type is not ideal for data flow analysis.
